src/autodiff.md

```diff
@@ -0,0 +1,108 @@
+
+A dual number  $x+\varepsilon x'$ consists of a primal/real part and a dual part dragging the derivative information. Every real number can be represented as $x+\varepsilon 0 $. The arithmetic is defined with the help of dummy symbol $\varepsilon$ by specifying that $\varepsilon^2=0$. Based on the above, the following set of rules can be easily derived.
+
+- $\left(x+\varepsilon x'\right)+\left(y+\varepsilon y'\right)=\left(x+y\right)+\varepsilon\left(x'+y'\right)$
+
+- $\left(x+\varepsilon x'\right)*\left(y+\varepsilon y'\right)=xy+\varepsilon\left(yx'+xy'\right)$
+
+- $f\left(x+\varepsilon x'\right)=f\left(x\right)+\varepsilon f'\left(x\right)x'$
+
+- $f\left(g \left(x+\varepsilon x'\right) \right)= f\left(g \left(x\right)+\varepsilon g'\left(x\right) x'\right) = f\left(g \left(x \right)\right)+\varepsilon f'\left(g \left(x \right)\right) g'\left(x\right) x'$
+
+
+# Example of AD differentiated function
+
+The following vector function, defined as lambda expression, has two parameters `kappa` and `load`. The input of the function `uu` is a vector with 4 components (the last one is not used in the output of the function), and the result is a vector `vres` of size 4.
+
+```c++
+//using lambda expression
+auto func = [](mfem::Vector& vparam, mfem::ad::ADVectorType& uu, mfem::ad::ADVectorType& vres) {
+   auto kappa = vparam[0]; //diffusion coefficient
+   auto load = vparam[1]; //volumetric influx
+
+   vres[0] = kappa * uu[0];
+   vres[1] = kappa * uu[1];
+   vres[2] = kappa * uu[2];
+   vres[3] = -load;
+};
+```
+The gradient of `vres` will be a matrix of size 4x4 and is computed with the help of the following object:
+```c++
+mfem::VectorFuncAutoDiff<4,4,2> fdr(func);
+```
+The first parameter in the above template specifies the length of the result, the second parameter the length of the input vector `uu`, and the third template parameter specifies the length of `vparam`. Once `fdr` is defined, the following statement computes the gradients:
+```c++
+fdr.QJacobian(param,state, grad_mat);
+```
+The input consists of parameters and a state vector, and the output is 4x4 `grad_mat` matrix.
+
+# Example of AD differentiated function using functors
+
+The following vector function, defined as a functor, has zero parameters. The input of the function `uu` is a vector with 6 components, and the result is a vector `rr` of size 3.
+
+```c++
+template<typename TDataType, typename TParamVector, typename TStateVector,
+         int residual_size, int state_size, int param_size>
+class ExampleResidual
+{
+public:
+    void operator ()(TParamVector& vparam, TStateVector& uu, TStateVector& rr)
+    {
+        rr[0]=sin(uu[0]+uu[1]+uu[2]);
+        rr[1]=cos(uu[1]+uu[2]+uu[3]);
+        rr[2]=tan(uu[2]+uu[3]+uu[4]+uu[5]);
+    }
+
+};
+```
+
+The gradient of `rr` will be a matrix of size 3x6 and is computed with the help of the following object:
```

Comment:
While it is implied, I think a note saying these class and function signatures must match this interface is needed.

Reply:
fixed